### PR TITLE
SBI Versioning 

### PR DESCRIPTION
**Pull request by Manas0606**

### CODE REVIEW

File: registration/registration-services/src/main/resources/props/mosip-application.properties [hunk 1]
```diff
@@ -1,7 +1,7 @@
 #1.1.2-rc2
 #Fri Jul 24 14:07:50 IST 2020
 environment=LOCAL
-mosip.hostname=dev.mosip.net
+mosip.hostname=api-internal.dst-dev.mosip.net
```
Comment (Rediket) [inline registration/registration-services/src/main/resources/props/mosip-application.properties]: @Manas0606 Remove this modification

File: registration/registration-client/src/main/java/io/mosip/registration/controller/GenericController.java [hunk 2]
```diff
@@ -557,7 +560,10 @@ public void changed(ObservableValue<? extends Number> observable, Number oldValu
 				//Hide continue button in preview page
 				next.setVisible(newScreenName.equals("AUTH") ? false : true);
 				authenticate.setVisible(newScreenName.equals("AUTH") ? true : false);
-
+				if(deviceSpecificationFactory.isVersionSupported==false||deviceSpecificationFactory.isAvailableDevice==false || deviceSpecificationFactory.deviceSize==0 ){
+					next.setDisable(newScreenName.equalsIgnoreCase("BiometricDetails"));
+				}
+                    //next.setDisable(newScreenName.equalsIgnoreCase("BiometricDetails"));
```
Comment (Rediket) [inline registration/registration-client/src/main/java/io/mosip/registration/controller/GenericController.java]: @Manas0606 line 566 has commented out code. Please remove this if it is not necessary.

File: registration/registration-client/src/main/java/io/mosip/registration/controller/device/GenericBiometricsController.java [hunk 3]
```diff
@@ -507,14 +507,88 @@ public void handle(WorkerStateEvent t) {
 		deviceSearchTask.setOnFailed(new EventHandler<WorkerStateEvent>() {
 			@Override
 			public void handle(WorkerStateEvent t) {
-				generateAlert(RegistrationConstants.ERROR, RegistrationUIConstants.getMessageLanguageSpecific(RegistrationUIConstants.NO_DEVICE_FOUND));
+				LOGGER.info("Entering handle method with WorkerStateEvent: " + t);
+
+				// Fetch map data for modalities
+				String mapDataFace = ApplicationContext.getStringValueFromApplicationMap("face");
```
Comment (Rediket) [inline registration/registration-client/src/main/java/io/mosip/registration/controller/device/GenericBiometricsController.java]: @Manas0606 please do not hard code string values such as "face", "iris", "unavailfinger", "fingerprint_slab_left" etc.... Put these values in the RegistrationConstants and load the values from there.

File: registration/registration-client/src/main/java/io/mosip/registration/controller/device/GenericBiometricsController.java [hunk 4]
```diff
@@ -507,14 +507,88 @@ public void handle(WorkerStateEvent t) {
...
+					{
+						message = "Attention: Unsupported finger device "+"Please use supported device";
+						LOGGER.error(message);
+					}
+				}
+				else if(currentModality.toString().toLowerCase().contains("iris_double")){
+					LOGGER.info("Current modality is iris.");
+					if(mapDataUnavailableIris!=null)
+					{
+						message = "Attention: Unsupported iris device "+"Please use supported device";
```
Comment (Rediket) [inline registration/registration-client/src/main/java/io/mosip/registration/controller/device/GenericBiometricsController.java]: @Manas0606 The error message has been hardcoded in English. You need to move the texts to the message.properties. Each language has its own message.properties file and this text needs to be translated and stored there. Then the translated text can be fetched using the line "_RegistrationUIConstants.getMessageLanguageSpecific(message)_"

File: registration/registration-client/src/main/java/io/mosip/registration/controller/device/GenericBiometricsController.java [hunk 5]
```diff
@@ -507,14 +507,88 @@ public void handle(WorkerStateEvent t) {
...
+				}
+				else if(currentModality.toString().toLowerCase().contains("iris_double")){
+					LOGGER.info("Current modality is iris.");
+					if(mapDataIris!=null)
+					{
+						message="Attention: Your current SBI version for the following modality "+mapDataIris+". Please upgrade your SBI to the latest version";
+						LOGGER.error(message);
+					}
+				}
+				if(currentModality.toString().toLowerCase().contains("face")){
```
Comment (Rediket) [inline registration/registration-client/src/main/java/io/mosip/registration/controller/device/GenericBiometricsController.java]: @Manas0606 The if-else clauses that display the error messages depending on whether or not the device is unavailable or SBI outdated are repeated. You can merge it into one code. For example:

```
else if(currentModality.toString().toLowerCase().contains("iris_double")){
					LOGGER.info("Current modality is iris.");
					if(mapDataIris!=null)
					{
						message="Attention: Your current SBI version for the following modality "+mapDataIris+". Please upgrade your SBI to the latest version";
						LOGGER.error(message);
					} else if (mapDataUnavailableIris!=null){
message = "Attention: Unsupported iris device "+"Please use supported device";
				LOGGER.error(message);	
				}
```

File: registration/registration-client/src/main/java/io/mosip/registration/controller/settings/impl/DeviceSettingsController.java [hunk 6]
```diff
@@ -1,530 +1,571 @@
...
+				mainGridPane.setNodeOrientation(NodeOrientation.RIGHT_TO_LEFT);
+			}
+			gridPane.add(mainGridPane, columnIndex, rowIndex);
+			rowIndex = (columnIndex == 1) ? (rowIndex + 1) : rowIndex;
+			columnIndex = (columnIndex == 1) ? 0 : (columnIndex + 1);
+		}
+		if (!message.isEmpty()) {
+			// Defer the alert generation until after the UI components have been rendered
+			String finalMessage = message;
+			Platform.runLater(() -> generateAlert(RegistrationConstants.ERROR, "Attention: " + finalMessage + " Please remove and rerun the application.\nNote: If the current version is unsupported then please Upgrade to latest SBI version or if device is unsupported then kindly switch to supported device"));
```
Comment (Rediket) [inline registration/registration-client/src/main/java/io/mosip/registration/controller/settings/impl/DeviceSettingsController.java]: @Manas0606 
1. Please move this text to message.properties
2. The grammar of this text is not valid. Please update it.

File: registration/registration-services/src/main/java/io/mosip/registration/mdm/service/impl/MosipDeviceSpecificationFactory.java [hunk 7]
```diff
@@ -213,15 +214,139 @@ private void initByPort(Integer availablePort) {
...
+		if (mdmBioDevices != null) {
+			for (MdmBioDevice device : mdmBioDevices) {
+				LOGGER.info("Processing device: " + device);
+				boolean isAvailable = true;
+				boolean isVersionAvailable = true;
+				try {
+					if (device.getDeviceProviderName() != null && device.getDeviceType() != null) {
+						String propertyValue = null;
+						switch (device.getDeviceType().toLowerCase()) {
+							case "face":
```
Comment (Rediket) [inline registration/registration-services/src/main/java/io/mosip/registration/mdm/service/impl/MosipDeviceSpecificationFactory.java]: move the the strings "face" "finger" etc ... into RegistrationConstants

File: registration/registration-services/src/main/java/io/mosip/registration/mdm/service/impl/MosipDeviceSpecificationFactory.java [hunk 8]
```diff
@@ -213,15 +214,139 @@ private void initByPort(Integer availablePort) {
...
+		if (value != null && !value.contains(data)) {
+			LOGGER.info("Appending to existing entry in map for key: " + key);
+			map.put(key, value + data + " " + device.getDeviceType() + " " + message);
+		} else {
+			LOGGER.info("Creating new entry in map for key: " + key);
+			map.put(key, data + " " + device.getDeviceType() + " " + message);
+		}
+	}
+
+	private boolean compareVersions(String deviceVersion, String configVersion) {
```
Comment (Rediket) [inline registration/registration-services/src/main/java/io/mosip/registration/mdm/service/impl/MosipDeviceSpecificationFactory.java]: @Manas0606  Can you explain more what this method does? It's supposed to take the SBI version and the list of unsupported versions, loop through each unsupported method and see if it matches the Current SBI version. Wouldnt the following method be simpler

```
 public static boolean isUnsupportedVersion(String SBIversion, List<String> unsupportedVersions) {
        // Convert SBIversion to a float
        float sbiVersionFloat = Float.parseFloat(SBIversion);

        // Iterate through the list of unsupported versions
        for (String version : unsupportedVersions) {
            // Convert each version in the list to a float
            float unsupportedVersionFloat = Float.parseFloat(version);

            // Compare the versions as floats
            if (sbiVersionFloat == unsupportedVersionFloat) {
                return true;
            }
        }

        // If no match is found, return false
        return false;
    }
```

File: registration/registration-services/src/main/java/io/mosip/registration/mdm/service/impl/MosipDeviceSpecificationFactory.java [hunk 9]
```diff
@@ -213,15 +214,139 @@ private void initByPort(Integer availablePort) {
...
+									versions.put(entry.getKey(), (List<String>) value);
+								}
+							}
+							LOGGER.debug("Versions map: " + versions);
+
+							String providerName = device.getDeviceProviderName().toUpperCase();
+							if (versions.containsKey(providerName)) {
+								List<String> unsupportedVersions = versions.get(providerName);
+								boolean isVersionUnSupported = unsupportedVersions.stream()
+										.anyMatch(unsupportedVersion -> compareVersions(device.getSerialVersion(), unsupportedVersion));
```
Comment (Rediket) [inline registration/registration-services/src/main/java/io/mosip/registration/mdm/service/impl/MosipDeviceSpecificationFactory.java]: @Manas0606 Is the SBI version stored here: device.getSerialVersion()? The SBI version is stored in the MdmDeviceInfo, serviceVersion variable

Comment (Manas0606) [reply registration/registration-services/src/main/java/io/mosip/registration/mdm/service/impl/MosipDeviceSpecificationFactory.java]: yes sbi version stores in device.getSerialVersion()

**Review by Rediket — APPROVED** [review]

**Pull request merged**